// ezhtml/index.js
/** Require local modules */
const elements = require('./elements');
const page = require(`./page`);

/** 
 * @module ezhtml
 * @author Rich Lowe
 * @copyright 2018 Rich Lowe
 * @description A simple library for programatically rendering HTML and client-side JavaScript using Node.js.
 */
module.exports.configAbbreviation = elements.configAbbreviation;
module.exports.configAddress = elements.configAddress;
module.exports.configAnchor = elements.configAnchor;
module.exports.configArea = elements.configArea;
module.exports.configArticle = elements.configArticle;
module.exports.configAside = elements.configAside;
module.exports.configAudio = elements.configAudio;
module.exports.configBase = elements.configBase;
module.exports.configBidirectionalIsolation = elements.configBidirectionalIsolation;
module.exports.configBidirectionalOverride = elements.configBidirectionalOverride;
module.exports.configBlockquote = elements.configBlockquote;
module.exports.configBody = elements.configBody;
module.exports.configBold = elements.configBold;
module.exports.configButton = elements.configButton;
module.exports.configCanvas = elements.configCanvas;
module.exports.configCaption = elements.configCaption;
module.exports.configChild = elements.configChild;
module.exports.configCitation = elements.configCitation;
module.exports.configCode = elements.configCode;
module.exports.configColumnGroup = elements.configColumnGroup;
module.exports.configColumn = elements.configColumn;
module.exports.configContainerElement = elements.configContainerElement;
module.exports.configContainer = elements.configContainer;
module.exports.configDataList = elements.configDataList;
module.exports.configData = elements.configData;
module.exports.configDefinitionDescription = elements.configDefinitionDescription;
module.exports.configDefinitionList = elements.configDefinitionList;
module.exports.configDefinitionTerm = elements.configDefinitionTerm;
module.exports.configDefinition = elements.configDefinition;
module.exports.configDeleted = elements.configDeleted;
module.exports.configDetails = elements.configDetails;
module.exports.configDialog = elements.configDialog;
module.exports.configDiv = elements.configDiv;
module.exports.configElement = elements.configElement;
module.exports.configEmbed = elements.configEmbed;
module.exports.configEmphasized = elements.configEmphasized;
module.exports.configFieldSet = elements.configFieldSet;
module.exports.configFigureCaption = elements.configFigureCaption;
module.exports.configFigure = elements.configFigure;
module.exports.configFooter = elements.configFooter;
module.exports.configForm = elements.configForm;
module.exports.configHead = elements.configHead;
module.exports.configHeader = elements.configHeader;
module.exports.configH1 = elements.configH1;
module.exports.configH2 = elements.configH2;
module.exports.configH3 = elements.configH3;
module.exports.configH4 = elements.configH4;
module.exports.configH5 = elements.configH5;
module.exports.configH6 = elements.configH6;
module.exports.configHorizontalRule = elements.configHorizontalRule;
module.exports.configHTML = elements.configHTML;
module.exports.configImage = elements.configImage;
module.exports.configInput = elements.configInput;
module.exports.configInserted = elements.configInserted;
module.exports.configItalic = elements.configItalic;
module.exports.configKeyboard = elements.configKeyboard;
module.exports.configLabel = elements.configLabel;
module.exports.configLegend = elements.configLegend;
module.exports.configLineBreak = elements.configLineBreak;
module.exports.configLink = elements.configLink;
module.exports.configListItem = elements.configListItem;
module.exports.configMain = elements.configMain;
module.exports.configMapElement = elements.configMapElement;
module.exports.configMark = elements.configMark;
module.exports.configMeta = elements.configMeta;
module.exports.configMeter = elements.configMeter;
module.exports.configMultimediaObject = elements.configMultimediaObject;
module.exports.configNav = elements.configNav;
module.exports.configNoScript = elements.configNoScript;
module.exports.configOptionGroup = elements.configOptionGroup;
module.exports.configOption = elements.configOption;
module.exports.configOrderedList = elements.configOrderedList;
module.exports.configOutput = elements.configOutput;
module.exports.configParagraph = elements.configParagraph;
module.exports.configParam = elements.configParam;
module.exports.configPicture = elements.configPicture;
module.exports.configPreformattedText = elements.configPreformattedText;
module.exports.configProgress = elements.configProgress;
module.exports.configQuotation = elements.configQuotation;
module.exports.configSample = elements.configSample;
module.exports.configScript = elements.configScript;
module.exports.configSection = elements.configSection;
module.exports.configSelect = elements.configSelect;
module.exports.configSmall = elements.configSmall;
module.exports.configSource = elements.configSource;
module.exports.configSpan = elements.configSpan;
module.exports.configStrikethrough = elements.configStrikethrough;
module.exports.configStrong = elements.configStrong;
module.exports.configStyle = elements.configStyle;
module.exports.configSubscript = elements.configSubscript;
module.exports.configSummary = elements.configSummary;
module.exports.configSuperscript = elements.configSuperscript;
module.exports.configSVG = elements.configSVG;
module.exports.configTableBody = elements.configTableBody;
module.exports.configTableData = elements.configTableData;
module.exports.configTableFooter = elements.configTableFooter;
module.exports.configTableHead = elements.configTableHead;
module.exports.configTableHeader = elements.configTableHeader;
module.exports.configTableRow = elements.configTableRow;
module.exports.configTable = elements.configTable;
module.exports.configTemplate = elements.configTemplate;
module.exports.configText = elements.configText;
module.exports.configTextArea = elements.configTextArea;
module.exports.configTime = elements.configTime;
module.exports.configTitle = elements.configTitle;
module.exports.configTrack = elements.configTrack;
module.exports.configUnderline = elements.configUnderline;
module.exports.configUnorderedList = elements.configUnorderedList;
module.exports.configVariable = elements.configVariable;
module.exports.configVideo = elements.configVideo;
module.exports.configWordBreakOpportunity = elements.configWordBreakOpportunity;

module.exports.Abbreviation = elements.Abbreviation;
module.exports.Address = elements.Address;
module.exports.Anchor = elements.Anchor;
module.exports.Area = elements.Area;
module.exports.Article = elements.Article;
module.exports.Aside = elements.Aside;
module.exports.Audio = elements.Audio;
module.exports.Base = elements.Base;
module.exports.BidirectionalIsolation = elements.BidirectionalIsolation;
module.exports.BidirectionalOverride = elements.BidirectionalOverride;
module.exports.Blockquote = elements.Blockquote;
module.exports.Body = elements.Body;
module.exports.Bold = elements.Bold;
module.exports.Button = elements.Button;
module.exports.Canvas = elements.Canvas;
module.exports.Caption = elements.Caption;
module.exports.Child = elements.Child;
module.exports.Citation = elements.Citation;
module.exports.Code = elements.Code;
module.exports.ColumnGroup = elements.ColumnGroup;
module.exports.Column = elements.Column;
module.exports.ContainerElement = elements.ContainerElement;
module.exports.Container = elements.Container;
module.exports.DataList = elements.DataList;
module.exports.DefinitionDescription = elements.DefinitionDescription;
module.exports.DefinitionList = elements.DefinitionList;
module.exports.DefinitionTerm = elements.DefinitionTerm;
module.exports.Definition = elements.Definition;
module.exports.Deleted = elements.Deleted;
module.exports.Details = elements.Details;
module.exports.Dialog = elements.Dialog;
module.exports.Div = elements.Div;
module.exports.Element = elements.Element;
module.exports.Embed = elements.Embed;
module.exports.Emphasized = elements.Emphasized;
module.exports.FieldSet = elements.FieldSet;
module.exports.FigureCaption = elements.FigureCaption;
module.exports.Figure = elements.Figure;
module.exports.Footer = elements.Footer;
module.exports.Form = elements.Form;
module.exports.Head = elements.Head;
module.exports.Header = elements.Header;
module.exports.HeadingGroup = elements.HeadingGroup;
module.exports.H1 = elements.H1;
module.exports.H2 = elements.H2;
module.exports.H3 = elements.H3;
module.exports.H4 = elements.H4;
module.exports.H5 = elements.H5;
module.exports.H6 = elements.H6;
module.exports.HorizontalRule = elements.HorizontalRule;
module.exports.HTML = elements.HTML;
module.exports.Image = elements.Image;
module.exports.Input = elements.Input;
module.exports.Inserted = elements.Inserted;
module.exports.Italic = elements.Italic;
module.exports.Keyboard = elements.Keyboard;
module.exports.Label = elements.Label;
module.exports.Legend = elements.Legend;
module.exports.LineBreak = elements.LineBreak;
module.exports.Link = elements.Link;
module.exports.ListItem = elements.ListItem;
module.exports.Main = elements.Main;
module.exports.MapElement = elements.MapElement;
module.exports.Mark = elements.Mark;
module.exports.MenuItem = elements.MenuItem;
module.exports.Menu = elements.Menu;
module.exports.Meta = elements.Meta;
module.exports.Meter = elements.Meter;
module.exports.MultimediaObject = elements.MultimediaObject;
module.exports.Nav = elements.Nav;
module.exports.NoScript = elements.NoScript;
module.exports.OptionGroup = elements.OptionGroup;
module.exports.Option = elements.Option;
module.exports.OrderedList = elements.OrderedList;
module.exports.Output = elements.Output;
module.exports.Page = page.Page;
module.exports.Paragraph = elements.Paragraph;
module.exports.Param = elements.Param;
module.exports.Picture = elements.Picture;
module.exports.PreformattedText = elements.PreformattedText;
module.exports.Progress = elements.Progress;
module.exports.Quotation = elements.Quotation;
module.exports.Sample = elements.Sample;
module.exports.Script = elements.Script;
module.exports.Section = elements.Section;
module.exports.Select = elements.Select;
module.exports.Small = elements.Small;
module.exports.Source = elements.Source;
module.exports.Span = elements.Span;
module.exports.Strikethrough = elements.Strikethrough;
module.exports.Strong = elements.Strong;
module.exports.Style = elements.Style;
module.exports.Subscript = elements.Subscript;
module.exports.Summary = elements.Summary;
module.exports.Superscript = elements.Superscript;
module.exports.SVG = elements.SVG;
module.exports.TableBody = elements.TableBody;
module.exports.TableData = elements.TableData;
module.exports.TableFooter = elements.TableFooter;
module.exports.TableHead = elements.TableHead;
module.exports.TableHeader = elements.TableHeader;
module.exports.TableRow = elements.TableRow;
module.exports.Table = elements.Table;
module.exports.Template = elements.Template;
module.exports.Text = elements.Text;
module.exports.TextArea = elements.TextArea;
module.exports.Time = elements.Time;
module.exports.Title = elements.Title;
module.exports.Track = elements.Track;
module.exports.Underline = elements.Underline;
module.exports.UnorderedList = elements.UnorderedList;
module.exports.Variable = elements.Variable;
module.exports.Video = elements.Video;
module.exports.WordBreakOpportunity = elements.WordBreakOpportunity;
