// ezhtml/elements/index.js
/** Require local modules */
const abbreviation = require('./abbreviation');
const address = require('./address');
const anchor = require('./anchor');
const area = require('./area');
const article = require('./article');
const aside = require('./aside');
const audio = require('./audio');
const base = require('./base');
const bidirectionalIsolation = require('./bidirectional-isolation');
const bidirectionalOverride = require('./bidirectional-override');
const blockquote = require('./blockquote');
const body = require('./body');
const bold = require('./bold');
const button = require('./button');
const canvas = require('./canvas');
const caption = require('./caption');
const child = require('./child');
const citation = require('./citation');
const code = require('./code');
const columnGroup = require('./column-group');
const column = require('./column');
const containerElement = require('./container-element');
const container = require('./container');
const dataList = require('./data-list');
const data = require('./data');
const definitionDescription = require('./definition-description');
const definitionList = require('./definition-list');
const definitionTerm = require('./definition-term');
const definition = require('./definition');
const deleted = require('./deleted');
const details = require('./details');
const dialog = require('./dialog');
const div = require('./div');
const element = require('./element');
const embed = require('./embed');
const emphasized = require('./emphasized');
const fieldset = require('./fieldset');
const figureCaption = require('./figure-caption');
const figure = require('./figure');
const footer = require('./footer');
const form = require('./form');
const head = require('./head');
const header = require('./header');
const headings = require('./headings');
const horizontalRule = require('./horizontal-rule');
const html = require('./html');
const image = require('./image');
const input = require('./input');
const inserted = require('./inserted');
const italic = require('./italic');
const keyboard = require('./keyboard');
const label = require('./label');
const legend = require('./legend');
const lineBreak = require('./line-break');
const link = require('./link');
const listItem = require('./list-item');
const main = require('./main');
const map = require('./map');
const mark = require('./mark');
const meta = require('./meta');
const meter = require('./meter');
const multimediaObject = require('./multimedia-object');
const nav = require('./nav');
const noScript = require('./no-script');
const optionGroup = require('./option-group');
const option = require('./option');
const orderedList = require('./ordered-list');
const output = require('./output');
const paragraph = require('./paragraph');
const param = require('./param');
const picture = require('./picture');
const preformattedText = require('./preformatted-text');
const progress = require('./progress');
const quotation = require('./quotation');
const sample = require('./sample');
const script = require('./script');
const section = require('./section');
const select = require('./select');
const small = require('./small');
const source = require('./source');
const span = require('./span');
const strikethrough = require('./strikethrough');
const strong = require('./strong');
const style = require('./style');
const subscript = require('./subscript');
const summary = require('./summary');
const superscript = require('./superscript');
const svg = require('./svg');
const tableBody = require('./table-body');
const tableData = require('./table-data');
const tableFooter = require('./table-footer');
const tableHead = require('./table-head');
const tableHeader = require('./table-header');
const tableRow = require('./table-row');
const table = require('./table');
const template = require('./template');
const textArea = require('./text-area');
const text = require('./text');
const time = require('./time');
const title = require('./title');
const track = require('./track');
const underline = require('./underline');
const unorderedList = require('./unordered-list');
const variable = require('./variable');
const video = require('./video');
const wordBreakOpportunity = require('./word-break-opportunity');

/** 
 * @module ezhtml
 * @author Rich Lowe
 * @copyright 2018 Rich Lowe
 * @description A simple library for programatically rendering HTML and client-side JavaScript using Node.js.
 */
module.exports.configAbbreviation = abbreviation.config;
module.exports.configAddress = address.config;
module.exports.configAnchor = anchor.config;
module.exports.configArea = area.config;
module.exports.configArticle = article.config;
module.exports.configAside = aside.config;
module.exports.configAudio = audio.config;
module.exports.configBase = base.config;
module.exports.configBidirectionalIsolation = bidirectionalIsolation.config;
module.exports.configBidirectionalOverride = bidirectionalOverride.config;
module.exports.configBlockquote = blockquote.config;
module.exports.configBody = body.config;
module.exports.configBold = bold.config;
module.exports.configButton = button.config;
module.exports.configCanvas = canvas.config;
module.exports.configCaption = caption.config;
module.exports.configChild = child.config;
module.exports.configCitation = citation.config;
module.exports.configCode = code.config;
module.exports.configColumnGroup = columnGroup.config;
module.exports.configColumn = column.config;
module.exports.configContainerElement = containerElement.config;
module.exports.configContainer = container.config;
module.exports.configDataList = dataList.config;
module.exports.configData = data.config;
module.exports.configDefinitionDescription = definitionDescription.config;
module.exports.configDefinitionList = definitionList.config;
module.exports.configDefinitionTerm = definitionTerm.config;
module.exports.configDefinition = definition.config;
module.exports.configDeleted = deleted.config;
module.exports.configDetails = details.config;
module.exports.configDialog = dialog.config;
module.exports.configDiv = div.config;
module.exports.configElement = element.config;
module.exports.configEmbed = embed.config;
module.exports.configEmphasized = emphasized.config;
module.exports.configFieldSet = fieldset.config;
module.exports.configFigureCaption = figureCaption.config;
module.exports.configFigure = figure.config;
module.exports.configFooter = footer.config;
module.exports.configForm = form.config;
module.exports.configHead = head.config;
module.exports.configHeader = header.config;
module.exports.configH1 = headings.config;
module.exports.configH2 = headings.config;
module.exports.configH3 = headings.config;
module.exports.configH4 = headings.config;
module.exports.configH5 = headings.config;
module.exports.configH6 = headings.config;
module.exports.configHorizontalRule = horizontalRule.config;
module.exports.configHTML = html.config;
module.exports.configImage = image.config;
module.exports.configInput = input.config;
module.exports.configInserted = inserted.config;
module.exports.configItalic = italic.config;
module.exports.configKeyboard = keyboard.config;
module.exports.configLabel = label.config;
module.exports.configLegend = legend.config;
module.exports.configLineBreak = lineBreak.config;
module.exports.configLink = link.config;
module.exports.configListItem = listItem.config;
module.exports.configMain = main.config;
module.exports.configMapElement = map.config;
module.exports.configMark = mark.config;
module.exports.configMeta = meta.config;
module.exports.configMeter = meter.config;
module.exports.configMultimediaObject = multimediaObject.config;
module.exports.configNav = nav.config;
module.exports.configNoScript = noScript.config;
module.exports.configOptionGroup = optionGroup.config;
module.exports.configOption = option.config;
module.exports.configOrderedList = orderedList.config;
module.exports.configOutput = output.config;
module.exports.configParagraph = paragraph.config;
module.exports.configParam = param.config;
module.exports.configPicture = picture.config;
module.exports.configPreformattedText = preformattedText.config;
module.exports.configProgress = progress.config;
module.exports.configQuotation = quotation.config;
module.exports.configSample = sample.config;
module.exports.configScript = script.config;
module.exports.configSection = section.config;
module.exports.configSelect = select.config;
module.exports.configSmall = small.config;
module.exports.configSource = source.config;
module.exports.configSpan = span.config;
module.exports.configStrikethrough = strikethrough.config;
module.exports.configStrong = strong.config;
module.exports.configStyle = style.config;
module.exports.configSubscript = subscript.config;
module.exports.configSummary = summary.config;
module.exports.configSuperscript = superscript.config;
module.exports.configSVG = svg.config;
module.exports.configTableBody = tableBody.config;
module.exports.configTableData = tableData.config;
module.exports.configTableFooter = tableFooter.config;
module.exports.configTableHead = tableHead.config;
module.exports.configTableHeader = tableHeader.config;
module.exports.configTableRow = tableRow.config;
module.exports.configTable = table.config;
module.exports.configTemplate = template.config;
module.exports.configText = text.config;
module.exports.configTextArea = textArea.config;
module.exports.configTime = time.config;
module.exports.configTitle = title.config;
module.exports.configTrack = track.config;
module.exports.configUnderline = underline.config;
module.exports.configUnorderedList = unorderedList.config;
module.exports.configVariable = variable.config;
module.exports.configVideo = video.config;
module.exports.configWordBreakOpportunity = wordBreakOpportunity.config;

module.exports.Abbreviation = abbreviation.Abbreviation;
module.exports.Address = address.Address;
module.exports.Anchor = anchor.Anchor;
module.exports.Area = area.Area;
module.exports.Article = article.Article;
module.exports.Aside = aside.Aside;
module.exports.Audio = audio.Audio;
module.exports.Base = base.Base;
module.exports.BidirectionalIsolation = bidirectionalIsolation.BidirectionalIsolation;
module.exports.BidirectionalOverride = bidirectionalOverride.BidirectionalOverride;
module.exports.Blockquote = blockquote.Blockquote;
module.exports.Body = body.Body;
module.exports.Bold = bold.Bold;
module.exports.Button = button.Button;
module.exports.Canvas = canvas.Canvas;
module.exports.Caption = caption.Caption;
module.exports.Child = child.Child;
module.exports.Citation = citation.Citation;
module.exports.Code = code.Code;
module.exports.ColumnGroup = columnGroup.ColumnGroup;
module.exports.Column = column.Column;
module.exports.ContainerElement = containerElement.ContainerElement;
module.exports.Container = container.Container;
module.exports.DataList = dataList.DataList;
module.exports.Data = data.Data;
module.exports.DefinitionDescription = definitionDescription.DefinitionDescription;
module.exports.DefinitionList = definitionList.DefinitionList;
module.exports.DefinitionTerm = definitionTerm.DefinitionTerm;
module.exports.Definition = definition.Definition;
module.exports.Deleted = deleted.Deleted;
module.exports.Details = details.Details;
module.exports.Dialog = dialog.Dialog;
module.exports.Div = div.Div;
module.exports.Element = element.Element;
module.exports.Embed = embed.Embed;
module.exports.Emphasized = emphasized.Emphasized;
module.exports.FieldSet = fieldset.FieldSet;
module.exports.FigureCaption = figureCaption.FigureCaption;
module.exports.Figure = figure.Figure;
module.exports.Footer = footer.Footer;
module.exports.Form = form.Form;
module.exports.Head = head.Head;
module.exports.Header = header.Header;
module.exports.H1 = headings.H1;
module.exports.H2 = headings.H2;
module.exports.H3 = headings.H3;
module.exports.H4 = headings.H4;
module.exports.H5 = headings.H5;
module.exports.H6 = headings.H6;
module.exports.HorizontalRule = horizontalRule.HorizontalRule;
module.exports.HTML = html.HTML;
module.exports.Image = image.Image;
module.exports.Input = input.Input;
module.exports.Inserted = inserted.Inserted;
module.exports.Italic = italic.Italic;
module.exports.Keyboard = keyboard.Keyboard;
module.exports.Label = label.Label;
module.exports.Legend = legend.Legend;
module.exports.LineBreak = lineBreak.LineBreak;
module.exports.Link = link.Link;
module.exports.ListItem = listItem.ListItem;
module.exports.Main = main.Main;
module.exports.MapElement = map.MapElement;
module.exports.Mark = mark.Mark;
module.exports.Meta = meta.Meta;
module.exports.Meter = meter.Meter;
module.exports.MultimediaObject = multimediaObject.MultimediaObject;
module.exports.Nav = nav.Nav;
module.exports.NoScript = noScript.NoScript;
module.exports.OptionGroup = optionGroup.OptionGroup;
module.exports.Option = option.Option;
module.exports.OrderedList = orderedList.OrderedList;
module.exports.Output = output.Output;
module.exports.Paragraph = paragraph.Paragraph;
module.exports.Param = param.Param;
module.exports.Picture = picture.Picture;
module.exports.PreformattedText = preformattedText.PreformattedText;
module.exports.Progress = progress.Progress;
module.exports.Quotation = quotation.Quotation;
module.exports.Sample = sample.Sample;
module.exports.Script = script.Script;
module.exports.Section = section.Section;
module.exports.Select = select.Select;
module.exports.Small = small.Small;
module.exports.Source = source.Source;
module.exports.Span = span.Span;
module.exports.Strikethrough = strikethrough.Strikethrough;
module.exports.Strong = strong.Strong;
module.exports.Style = style.Style;
module.exports.Subscript = subscript.Subscript;
module.exports.Summary = summary.Summary;
module.exports.Superscript = superscript.Superscript;
module.exports.SVG = svg.SVG;
module.exports.TableBody = tableBody.TableBody;
module.exports.TableData = tableData.TableData;
module.exports.TableFooter = tableFooter.TableFooter;
module.exports.TableHead = tableHead.TableHead;
module.exports.TableHeader = tableHeader.TableHeader;
module.exports.TableRow = tableRow.TableRow;
module.exports.Table = table.Table;
module.exports.Template = template.Template;
module.exports.Text = text.Text;
module.exports.TextArea = textArea.TextArea;
module.exports.Time = time.Time;
module.exports.Title = title.Title;
module.exports.Track = track.Track;
module.exports.Underline = underline.Underline;
module.exports.UnorderedList = unorderedList.UnorderedList;
module.exports.Variable = variable.Variable;
module.exports.Video = video.Video;
module.exports.WordBreakOpportunity = wordBreakOpportunity.WordBreakOpportunity;
